### ISSUE TITLE
Problems with variable handling in inventory files

### ISSUE THREAD

**Issue by willharris** [ansible/ansible#14581]
##### Issue Type:

Bug Report
##### Ansible Version:

```
ansible 2.0.0.2
  config file = /Users/harris/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

N/A
##### Environment:

N/A
##### Summary:

There are a couple of issues involving the handling of variables defined in a `[group:vars]` block in an inventory file with respect to variables defined in the `group_vars` directory:
- inconsistency between 1.9.4 and 2.0.0.2 when overriding variables defined on child groups.
- inability to override `all` vars using such a block in 2.0.0.2.

In addition, it is not clear from the documentation about variable scope at

http://docs.ansible.com/ansible/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

where the vars placed in the `group_vars` directory are expected to be considered.
##### Steps To Reproduce:

I've made a GitHub repo to demonstrate the problem:

https://github.com/willharris/ansible-variables-test

```
ansible-playbook test.yml -i test_inv
```

This step, using the HEAD of my repo, shows the simplest form of the problem - the inability to even override an `all` variable in the inventory.

I have two other commits in the repo that show other confusing aspects of the problem: depending on whether the you override variables on a child group or a parent group, the behaviour was different in 1.9.4 compared to 2.0.0.2.
##### Expected Results:

I would expect the same variable value as was possible with 1.9.4:

```
ok: [inventory-simple_grp-value.test.com] => {
    "var": {
        "test_var": "inventory-simple_grp-value"
    }
}
```

The variable value should be "inventory-simple_grp-value".

I would expect that a variable defined in the inventory file would have precedence over all variables defined within the `group_vars` directory, as I would expect those to be more like defaults, whereas the inventory is a concrete set of "real-world" machines.
##### Actual Results:

The variable value from `all` is taken.

```
ok: [inventory-simple_grp-value.test.com] => {
    "test_var": "all-group-vars-value"
}
```

**Comment by coppit** [comment]
What are the prospects for getting this fixed some time soon? This is preventing us from upgrading to Ansible 2. Thanks!

**Comment by jimi-c** [comment]
I'm looking to get it in for 2.1.

**Comment by coppit** [comment]
Ug. And now the EPEL repos have switched to 2.0, forcing us to implement a workaround to continue using 1.9.

Thanks for working on this. Looking forward to being able to enjoy the 2.0 goodness. :)

**Comment by jimi-c** [comment]
So the issue where `all` groups are not overridden by things in `[some_group:vars]` was fixed yesterday. Looking into the other reported inconsistency to see if it's still a problem.

**Comment by jimi-c** [comment]
@willharris / @coppit it looks like my fix yesterday for the `all` precedence issue fixed the above as well. When I run the provided example, the output I get is:

<pre>
TASK [debug] *******************************************************************
task path: /root/testing/14581/ansible-variables-test/test.yml:5
ok: [inventory-simple_grp-value.test.com] => {
    "test_var": "inventory-simple_grp-value"
}
PLAY RECAP *********************************************************************
inventory-simple_grp-value.test.com : ok=1    changed=0    unreachable=0    failed=0 
</pre>

^ which matches the 1.9.4 output.

Based on the above, I'm going to go ahead and close this now. The fix will be in 2.1.0.

If you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
